apt: include apt preferences (e.g. pinning) when selecting packages

phemmer opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Factor in apt's preferences files when determining candidate packages to install.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #77969

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
apt

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

phemmer committed:
apt: include apt preferences (e.g. pinning) when selecting packages

Fixes #77969

s-hertel reviewed (approved):
Thanks for the quick fix.

Would you be able to complete this with tests? Feel free to modify those that are here if it's convenient: https://github.com/ansible/ansible/pull/78319/files.

phemmer commented:
Sure, question then: What do we expect the behavior to be if someone has pinned version 1.0, and they do `>=1.1`? Should it install 1.1, or should it install nothing?

bcoca commented:
it should 'fail' and note the conflict so user can either override the pinning or remove the version specified

noonedeadpunk commented on lib/ansible/modules/apt.py:
```diff
@@ -687,23 +711,21 @@ def install(m, pkgspec, cache, upgrade=False, default_release=None,
-        if (not installed and not only_upgrade) or (installed and not installed_version) or (upgrade and version_installable):
-            if version_installable or version:
-                pkg_list.append("'%s=%s'" % (name, version_installable or version))
+        if (not installed and not only_upgrade) or (installed and not installed_version and version_installable) or (upgrade and version_installable):
```
I'm not sure I read this logic properly, or well, fully see all cases when version will be specified for the package installation (either because of my skills or overall code readability), so having some comment about intended behavior would be useful here IMO.

noonedeadpunk reviewed (commented):
From what I see, ansible now is not providing version to that apt when it's not required. And pinning usecases I tried seems working now as well.

phemmer commented:
> From what I see, ansible now is not providing version to that apt when it's not required.

No, it's the opposite. It's now passing the version more. There were a few corner cases where it wasn't passing the version.

phemmer commented:
> @bcoca 
> it should 'fail' and note the conflict so user can either override the pinning or remove the version specified

Ok, doing this required a little more change than just the original 2 lines. However it also fixed a few cases where if you asked the module to install a version, but no version matched, it would silently do nothing (no change, but also no error).

Note that the way I addressed this was though this: https://github.com/ansible/ansible/pull/78327/files#diff-821fcd0036eed7308025fa9e825de90fb23fd6ffe6d86e82db56802c5fddc0f7R715-R718. However for virtual packages where we can't figure out what real package to install, and where we pass `apt-get` a package name without version specifier, it's still possible to get a `0 upgraded, 0 newly installed` result where we have no change and no error. It might be prudent to convert `0 upgraded, 0 newly installed` into an error. However this is not new behavior, and it would seem has been this way for ages. So could also just leave it.

bcoca commented:
@noonedeadpunk debug will log the actual commands to the log of the remote system (syslog/journal)

@phemmer I'm ok with splitting that into it's own PR

phemmer commented:
> > No, it's the opposite. It's now passing the version more. There were a few corner cases where it wasn't passing the version.
> 
> well, even with debug ansible not showing actual command that is ran, but according to output I'd say that version is not provided here:

That's a virtual package. When the module can't figure out what real package is to satisfy a virtual package, it stops trying and lets `apt-get` figure it out.

> But why explicitly add version to the package, when operator haven't asked for it?

It doesn't matter if the operator has asked for it. The operator asked for a package, why should it matter if ansible figures out the version or if apt-get figures out the version?

> What motivation is behind that and why not just trust system dependency resolution?

Because ansible needs to know what the version candidates are so that it can properly do change detection. Once ansible knows the version, there is no harm in using the version. Yes we could add code to explicitly strip the version off when the user didn't specify it, but there is no point to doing so.

> It's more common usecase not to specify version to the package, and I believe it should execute faster as well, as some part of the code with dependency resolution can be just skipped?

Because as mentioned, ansible needs to figure out whether it should do something before actually doing it. In the process of checking whether there is a package that can be installed, it also obtains the version information. So determining the version and passing it to `apt-get` does not add additional overhead.

s-hertel commented:
@noonedeadpunk The dependency resolution would need to happen regardless - if no version is specified, the module checks for upgrades. This just also identifies which upgrade - which, you're right, is not strictly necessary if no version or release is specified since the outcome should be identical. But this seems simpler to me than special-casing no version + no release. It also has the perk of allowing check mode + diff mode to reflect what would be installed.

noonedeadpunk commented:
> debug will log the actual commands to the log of the remote system (syslog/journal)

yeah, eventually it's in history.log and command had version specified: `Commandline: /usr/bin/apt-get -y -o Dpkg::Options::=--force-confdef -o Dpkg::Options::=--force-confold install erlang-base=1:22.2.7+dfsg-1`


Ok, thanks for explanation.
I believe you're right, just setting version feels a bit fragile for me at least and feels like having more maintenance cost. But since it's needed and done anyway, likely it doesn't matter indeed, as long as it works.

phemmer commented:
Hold on merging. I think I see an issue around the use of `only_upgrade`. Will add a test case.

phemmer commented:
> Hold on merging. I think I see an issue around the use of `only_upgrade`. Will add a test case.

Was actually no issue. So good to go.

s-hertel commented on lib/ansible/modules/apt.py:
```diff
@@ -480,12 +480,17 @@ def package_version_compare(version, other_version):
-        policy.create_pin('Version', pkgname, version, 991)
+        # Installing a specific version from command line overrides all pinning
+        # We don't mimmic this exactly, but instead set a priority which is higher than all APT built-in pin priorities.
+        policy.create_pin('Version', pkgname, version, 1001)
```
The module documentation should include the priority for release and = now that other pins are taken into account.

s-hertel commented on changelogs/fragments/77969-apt-preferences.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - apt - fix package selection to include /etc/apt/preferences(.d)
```
This is a bugfix and should be backported to 2.13.

```suggestion
bugfixes:
  - apt - fix package selection to include /etc/apt/preferences(.d) (https://github.com/ansible/ansible/issues/77969)
```

s-hertel reviewed (commented):
Changes still lgtm

s-hertel commented:
rebuild_merge
